stores/sql/metrics.go

```diff
@@ -0,0 +1,203 @@
+	}
+	defer insertStmt.Close()
+
+	for _, metric := range metrics {
+		res, err := insertStmt.Exec(ctx,
+			time.Now().UTC(),
+			UnixTimeMS(metric.Timestamp),
+			FileContractID(metric.ContractID),
+			PublicKey(metric.HostKey),
+			metric.HostVersion,
+			Unsigned64(metric.Pruned),
+			Unsigned64(metric.Remaining),
+			metric.Duration,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to insert contract prune metric: %w", err)
+		} else if n, err := res.RowsAffected(); err != nil {
+			return fmt.Errorf("failed to get rows affected: %w", err)
+		} else if n == 0 {
+			return fmt.Errorf("failed to insert contract prune metric: no rows affected")
+		}
+	}
+
+	return nil
+}
+
+func RecordContractSetMetric(ctx context.Context, tx sql.Tx, metrics ...api.ContractSetMetric) error {
+	insertStmt, err := tx.Prepare(ctx, "INSERT INTO contract_sets (created_at, timestamp, name, contracts) VALUES (?, ?, ?, ?)")
+	if err != nil {
+		return fmt.Errorf("failed to prepare statement to insert contract set metric: %w", err)
+	}
+	defer insertStmt.Close()
+
+	for _, metric := range metrics {
+		res, err := insertStmt.Exec(ctx,
+			time.Now().UTC(),
+			UnixTimeMS(metric.Timestamp),
+			metric.Name,
+			metric.Contracts,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to insert contract set metric: %w", err)
+		} else if n, err := res.RowsAffected(); err != nil {
+			return fmt.Errorf("failed to get rows affected: %w", err)
+		} else if n == 0 {
+			return fmt.Errorf("failed to insert contract set metric: no rows affected")
+		}
+	}
+
+	return nil
+}
+
+func queryPeriods(ctx context.Context, tx sql.Tx, start time.Time, n uint64, interval time.Duration, opts interface{}) (*sql.LoggedRows, error) {
+	if n > api.MetricMaxIntervals {
+		return nil, api.ErrMaxIntervalsExceeded
+	}
+	params := []interface{}{
+		UnixTimeMS(start),
+		interval.Milliseconds(),
+		UnixTimeMS(start.Add(time.Duration(n) * interval)),
+		interval.Milliseconds(),
+		interval.Milliseconds(),
+	}
+
+	where, err := whereClauseFromQueryOpts(opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build where clause: %w", err)
+	} else if len(where.params) > 0 {
+		params = append(params, where.params...)
+	}
+
+	return tx.Query(ctx, fmt.Sprintf(`
+		WITH RECURSIVE periods AS (
+			SELECT ? AS period_start
+			UNION ALL
+			SELECT period_start + ?
+			FROM periods
+			WHERE period_start < ? - ?
+		)
+		SELECT %s.* FROM %s
+		INNER JOIN (
+		SELECT
+			p.period_start as Period,
+			MIN(obj.id) AS id
+		FROM
+			periods p
+		INNER JOIN
+			%s obj ON obj.timestamp >= p.period_start AND obj.timestamp < p.period_start + ?
+		WHERE %s
+		GROUP BY
+			p.period_start
+		) i ON %s.id = i.id ORDER BY Period ASC
+	`, where.table, where.table, where.table, where.query, where.table), params...)
+}
+
+type whereClause struct {
+	table  string
+	query  string
+	params []interface{}
+}
+
+func whereClauseFromQueryOpts(opts interface{}) (where whereClause, _ error) {
```

Comment:
If this is only ever called from `queryPeriods` I'd inline it to avoid the cognitive load of separating this logic from the context it is used in and to avoid the new `whereClause` type.